### PR TITLE
Add TextObject scale functions

### PR DESCRIPTION
**Pull request by Wend1go**
This adds scale functions to TextObjects.
In the first step just for the Pixi renderer.
The Tween extension can even scale texts out of the box without further modifications. :)
The scale functions still need to be added to the IDE as standalone conditions, actions and expressions.

@4ian would you mind explaining your setup for developing the C++ part?
I'd be interested in the IDE setup you are using and how you do the debugging.
There is more functionality in the C++ part that is missing and I like to add it. Like opacity and object colour of PanelSprites and TiledSprites.

### CODE REVIEW

**Comment by Wend1go** [comment]
Looks like semaphoreci has issues with downloading libGD.js too.

**Comment by 4ian** [comment]
> @4ian would you mind explaining your setup for developing the C++ part?
> I'd be interested in the IDE setup you are using and how you do the debugging.

Sure, check out this README: https://github.com/4ian/GDevelop/tree/master/GDevelop.js
You have to get the Emscripten compiler ("C++ to JavaScript" compiler), precisely as explained in the README. Once you have it, you run the compilation. If it works, you have a version of libGD.js that is built from your computer :)

Are you on Linux though? I never tried to do this compilation on Windows.

Continuous integration servers are building libGD.js too, you can check [the configuration here](https://github.com/4ian/GDevelop/blob/master/.circleci/config.yml#L16-L22).

It's basically what is written in the README:

https://github.com/4ian/GDevelop/blob/d290c98e28dfa580e2153b64c2c1b592be20ed6e/.circleci/config.yml#L16-L22
https://github.com/4ian/GDevelop/blob/d290c98e28dfa580e2153b64c2c1b592be20ed6e/.circleci/config.yml#L54-L57

**Comment by Wend1go** [comment]
This worked like a charm :+1:

**Comment by 4ian** [comment]
Awesome! You can start trying to do changes in Cpp extensions (for example, the text extension is created here: https://github.com/4ian/GDevelop/blob/master/Extensions/TextObject/Extension.cpp and associated JS functions are plugged in here: https://github.com/4ian/GDevelop/blob/master/Extensions/TextObject/JsExtension.cpp), save and re-run `npm run build` (and `npm test`) after any change.

File: Extensions/TextObject/Extension.cpp [hunk 1]
```diff
@@ -108,6 +108,68 @@ void DeclareTextObjectExtension(gd::PlatformExtension& extension) {
...
+      .AddParameter("relationalOperator", _("Sign of the test"))
+      .AddParameter("expression", _("x-scale to test"))
+      .SetFunctionName("GetScaleX")
+      .SetManipulatedType("number")
+      .SetIncludeFile("TextObject/TextObject.h");
+
+  obj.AddAction(
+         "ScaleX",
+         _("Scale on X axis"),
+         _("Change the scale of the text on the X axis. (1.0 is default)"),
```
Comment (4ian) [inline Extensions/TextObject/Extension.cpp]: nitpicking: `Change the scale of the text on the X axis (default scale is 1)`

**Comment by Wend1go** [comment]
I tried to add a scale action to scale both axes at one but I get the error message below.

grafik

I'm not sure if I completely understood how the JsExtension.cpp works. Do I need to assign a getter?

JsExtension.cpp:
```
    GetAllActionsForObject("TextObject::Text")["TextObject::Scale"]
        .SetFunctionName("setScale")
        .SetIncludeFile("Extensions/TextObject/textruntimeobject.js");
```
Extension.cpp:
```
  obj.AddAction(
         "Scale",
         _("Scale"),
         _("Change the scale of the specified object (default scale is 1)"),
         _("Change scale of _PARAM0_ to _PARAM1__PARAM2_"),
         _("Scale"),
         "res/actions/scale24.png",
         "res/actions/scale.png")

      .AddParameter("object", _("Object"), "Text")
      .AddParameter("operator", _("Modification's sign"))
      .AddParameter("expression", _("Value"))
      .SetFunctionName("SetScale")
      .SetManipulatedType("number")
      .SetIncludeFile("TextObject/TextObject.h");
```

textruntimeobject.js
```
/**
 * Set the text object scale.
 * @param {number} newScale The new scale for the text object.
 */
gdjs.TextRuntimeObject.prototype.setScale = function(newScale) {
    this._scaleX = newScale;
    this._scaleY = newScale;
    this._renderer.setScale(newScale);
};
```

textruntimeobject-pixi-renderer.js
```
/**
 * Set the text object scale.
 * @param {number} newScale The new scale for the text object.
 */
gdjs.TextRuntimeObjectPixiRenderer.prototype.setScale = function(newScale) {
    this._text.scale.x = newScale;
    this._text.scale.y = newScale;
};
```

If I add `.SetGetter("getScale")` in the JsExtension.cpp the error vanishes but since `scale` holds the values for x and y, I can't provide those as a single value.

**Comment by 4ian** [comment]
Two issues:

- As you highlighted, you need to use SetGetter("getScale")
- Beside that, as you also said, `getScale` returns an object, not a number. (The issue also exists in JavaScript, you can't "add" an object and a number would make no sense).

The solution I opted for, for Sprites, is to return the average:
https://github.com/4ian/GDevelop/blob/1c7d11a302cb3d4054cd52af7e38c600ebd194c8/GDJS/Runtime/spriteruntimeobject.js#L855-L857

So that you can still use the +, -, /, * operators to set the scale on both X and Y at once, knowing that it will work if the scale and X and Y are the same. If they are different, you'll get the average - so that it still works but the user should understand that anytime they are using the scale action, the scale on X and Y are supposed to be the same.

File: Extensions/TextObject/textruntimeobject-pixi-renderer.js [hunk 2]
```diff
@@ -76,3 +76,49 @@ gdjs.TextRuntimeObjectPixiRenderer.prototype.getWidth = function() {
 gdjs.TextRuntimeObjectPixiRenderer.prototype.getHeight = function() {
     return this._text.height;
 };
+
+/**
+ * Get scale of the text.
+ */
+gdjs.TextRuntimeObjectPixiRenderer.prototype.getScale = function() {
```
Comment (4ian) [inline Extensions/TextObject/textruntimeobject-pixi-renderer.js]: This is not used right? Can we remove this to decrease the API surface and avoid potential confusions? :)

Comment (Wend1go) [reply Extensions/TextObject/textruntimeobject-pixi-renderer.js]: done

**Comment by Wend1go** [comment]
Should be ready to merge now, I guess.

**Comment by 4ian** [comment]
Looks good, let me check one last time and merge this :)

**Comment by 4ian** [comment]
I've updated the sentences of the action to be consistent with the rest of actions like this in GD.
Thanks for the PR! Will be useful for doing nice effects with texts :)

**Pull request merged**